<center> <h1> DS 3000 - HW 1</h1> </center>

<center> <h3> Insert your full name here by double-clicking </h3></center>

## What to do

* Carefully read the description below.
* **As you write your code, please include comments, when necessary**
  * When you are doing a complicated thing, for instance.
  * This is not needed for simple things.
* Make sure you execute your code for each question and **compare** the output to the sample output provided, if any.
* **Refrain** from using advanced Python capabilities that were not covered in class.
* **When in doubt, ASK!** Post a question on the Piazza discussion forum for this homework assignment.
* Complete the assignment on your own! **No collaboration is allowed.**

## What to submit
* When you are done with the assignment, save it.
* If using Colab, download it as a **Jupyter Notebook** file (DS3000_HW1.ipynb)
* If using Anacoda, upload your completed Notebook to Colab (https://colab.research.google.com/notebooks/), as shown in the video.
* Get a shareable link to your Notebook on Colab
    * Go to Share near the top-right corner of the screen.
    * <font color='red'> **Make sure you select Anyone with the Link Can View** </font>
    * Copy the link
* Go to Canvas
* **Upload the physical copy of Jupyter Notebook through the Canvas assignment link.**
* **Paste the link to your Notebook hosted on Colab in the Comments field.**
* Both are required!
* You can submit multiple attempts. The last attempt will be graded.
* Submissions failing to follow these instructions will receive a grade of 0.

# <center> Part 1: Data Avengers... Assemble! </center>

<center><img src = "https://vignette.wikia.nocookie.net/marvelcinematicuniverse/images/0/0a/Nick_Fury_Textless_AoU_Poster.jpg" width = 150></center>

Nick Fury is waiting for a assignment from the council. While waiting, he decides to implement the popular dice game 'Craps' in Python. You know, why not? :) Your task is to help Fury with this.

**The rules of the game are as follows:**

* You roll two six-sided dice, each with faces containing one, two, three, four, five and six spots, respectively. 
* When the dice come to rest, the sum of the spots on the two upward faces is calculated. 
    * If the sum is 7 or 11 on the first roll, you win. 
    * If the sum is 2, 3 or 12 on the first roll (called “craps”), you lose (i.e., the “house” wins). 
    * If the sum is 4, 5, 6, 8, 9 or 10 on the first roll, that sum becomes your “point.” 
* To win, you must continue rolling the dice until you “make your point” (i.e., roll that same point value). In other words, if the sum of your new roll is equal to your point, you win.
* You lose by rolling a 7 before making your point.
* Thus, the game continues until you win or lose.

**The requirements are as follows:**

* Your script should be designed to play 10000 games of craps. In the end, a summary of the results (number of wins/losses) will be presented as described below.
* Use a **wins dictionary** to keep track of the number of games won for a particular number of rolls (remember the game continues until you win, meaning a single game might contain many rolls). For instance, a typical key-value pair in the **wins dictionary** may be 
    * **5: 3000**
    * indicating that 3000 games were won on the 5th roll.
    
* Likewise, use a **losses dictionary** to keep track of the number of games lost for a particular number of rolls.
* As the simulation proceeds, keep updating the dictionaries.
* The dictionaries should **store the number of wins/losses for up to 20 rolls**. After 20 rolls, the game should consider those as wins/losses after the 20th roll. You should use a key of 21 to represent these in your dictionary. Thus the sizes of the wins and losses dictionaries should be 21.


When in doubt, take a break, come back, and read the description again.

## Q1 (2 points)
Write a **function**, called *roll_dice*, that randomly rolls two dice and return their face values as a **tuple** (int). Refer to the sample output below.

In [ ]:
def roll_dice():
   #answer goes here

In [ ]:
roll_dice()

(6, 3)

## Q2 (8 points)
Write a function, *play_craps*, that plays **10000 games of craps** based on the rules and requirement specified above. The  value of the number of games should be modifiable. Thus, your function should accept the number_of_games variable as a paramater instead of using the hard-coded value of 10000.

This script should begin by making the first roll (using the roll_dice function you just defined). Your script should keep track of **wins and losses** and store their counts for each roll in their corresponding dictionaries. The sample output below displays the contents of the wins and losses dictionaries after this code snippet is executed.

In [ ]:
#for your convenience the dictionaries are already defined and you should use them.:
wins = {}
losses = {}

def play_craps(number_of_games):
    
    #answer goes here

In [ ]:
play_craps(10000)

In [ ]:
wins # 1:2267 indicates 2267 games were won on the 1st roll
#note that the indices won't be ordered sequentially

{3: 544,
 2: 760,
 7: 143,
 5: 281,
 1: 2267,
 6: 212,
 4: 386,
 9: 76,
 10: 59,
 8: 107,
 14: 19,
 15: 10,
 11: 36,
 16: 9,
 13: 18,
 17: 3,
 20: 2,
 21: 7,
 12: 23,
 19: 2,
 18: 3}

In [ ]:
losses # 3:776 indicates 776 games were lost on the 3rd roll

{1: 1089,
 3: 776,
 2: 1093,
 5: 431,
 4: 548,
 9: 120,
 8: 159,
 21: 7,
 10: 63,
 12: 41,
 11: 62,
 6: 298,
 7: 237,
 14: 22,
 15: 20,
 19: 5,
 18: 10,
 13: 27,
 17: 10,
 16: 12,
 20: 3}

## Q3 (3 points)
Write a function that displays the results of the game based on the previous step. Your code should present the number of the wins and losses for the number of rolls values stored in the dictionaries, as shown below. Avoid hard-coding the values.

At the end of the output, the function should display the probability of winning across all the games (which can be calculated as the proportion of the number of wins to the total number of games).

You can use the escape characters "\n" and "\t" to format your output.

In [ ]:
def display_results():
    
    #answer goes here

In [ ]:
display_results()

ROLLS	WINS	LOSSES
1 	 2267 	 1089
2 	 760 	 1093
3 	 544 	 776
4 	 386 	 548
5 	 281 	 431
6 	 212 	 298
7 	 143 	 237
8 	 107 	 159
9 	 76 	 120
10 	 59 	 63
11 	 36 	 62
12 	 23 	 41
13 	 18 	 27
14 	 19 	 22
15 	 10 	 20
16 	 9 	 12
17 	 3 	 10
18 	 3 	 10
19 	 2 	 5
20 	 2 	 3
20+	 7 	 7

The probability of winning is 4967 / 10000 = 0.4967


<hr>

**Computer chimes...**
**Computer chimes...**
**Computer chimes...**
**Computer chimes...**
**Computer chimes...**

**Nick Fury**: What is going on??? <br>
**Maria Hill**: Colonel, look at this! Someone sent this! 

(Fury looks at the screen and sees a long sequence of 0s and 1s!)

**Nick Fury**: It's a string in binary!


In [ ]:
received_message = "01100101011000110110010001100100011000110110010101100011011001000110010001100110011001010110011001100001"

**Nick Fury**: We gotta decode this right NOW! This is just insane!

<hr>
Note: This is a binary string with each 8-bit sequence representing a lower case character.


## Q4 (2 points)
Write a one-line function 'split_message' that would take in the entire string and split it into a list of strings, each 8-character long, and return that list.

For example: <br>
"0011100101100101" would split into ["00111001", "01100101"]


In [ ]:
def split_message(message):
    #answer goes here

In [ ]:
clean_message = split_message(received_message)
clean_message

['01100101',
 '01100011',
 '01100100',
 '01100100',
 '01100011',
 '01100101',
 '01100011',
 '01100100',
 '01100100',
 '01100110',
 '01100101',
 '01100110',
 '01100001']

## Q5 (2 points)

You will now need to decode the cleaned message. For this purpose, first write a function that retuns a dictionary of lowercase letters and their 8-bit binary codes, which you will need in the next step. The dictionary would contain 8-bit codes as keys and their corresponding character as value, as shown in the sample output.

- Use the helper function 'decode_character' to convert a character into its 8-bit binary equivalent. 
- The helper function takes in a character and returns the 8-bit representation
- You will need to generate a dictionary of letter characters and their binary code before you can decode the message. This is what you are doing in this step.

The dictionary would look something like this:
dict_bin_text = {
  '01100001': 'a',
  '01100010': 'b',
  ...
}

**Hint:** You will need to use two new built-in functions: ord() and chr(). 

* ord('a') = 97 and chr(97) = 'a'. 
* In-built 'ord' function would return ASCII value of the given character. Whereas 'chr' would return character given the ASCII value.

Try <br> 
<i><center>for i in range('a', 'z')</center></i><br> Error? You know what to do!

<hr>
Note: There are only lowercase letters in the message. To iterate over 'a' -> 'z', you can iterate over 97 -> 123. Pause for a moment and think about why that is the case.

In [ ]:
#use this helper function when decoding characters
#it returns the ASCII code for the character
def decode_character(x):
    return format(x, '08b')

In [ ]:
def make_dict_bin_text():
   #answer goes here

In [ ]:
dict_bin_text = make_dict_bin_text()
dict_bin_text

{'01100001': 'a',
 '01100010': 'b',
 '01100011': 'c',
 '01100100': 'd',
 '01100101': 'e',
 '01100110': 'f',
 '01100111': 'g',
 '01101000': 'h',
 '01101001': 'i',
 '01101010': 'j',
 '01101011': 'k',
 '01101100': 'l',
 '01101101': 'm',
 '01101110': 'n',
 '01101111': 'o',
 '01110000': 'p',
 '01110001': 'q',
 '01110010': 'r',
 '01110011': 's',
 '01110100': 't',
 '01110101': 'u',
 '01110110': 'v',
 '01110111': 'w',
 '01111000': 'x',
 '01111001': 'y',
 '01111010': 'z'}

## Q6 (2 points)

Now that you have this dictionary, you can go ahead and decode the message stored in "clean_message". For this purpose, write a function, called "decode", that takes in the message and returns a string that decodes and **joins** the individual characters present in the "clean_message" variable. For decoding, use the dict_bin_text from the previous step. Refer to the sample output.

**Note:** The length of the decoded string should be the same as the length of the clean_message list.

In [ ]:
def decode(clean_message):
    #answer goes here

In [ ]:
decoded_message = decode(clean_message)
decoded_message

<hr>

*** the decoded message is hidden intentionally; you'll need to figure out which file you should use yourself. :) ***

<hr>
**Nick Fury**: What is this?..some name?..some file?..some ID? What is this? 

**Maria**: I've just searched it in our database, Colonel! There is a file matching this name in our data folder. Go ahead and open it.

**Refer to the data folder attached to the assignment; make sure it's saved under the same directory as this notebook file**

**To all the readers**:
*Turns out Nick Fury got his hands on a secret file with statistics of all the super heros (yes, super villans as well!) in the universe. He wants to know where his army stands, which will help him prepare to keep the universe safe. The council won't allow Fury to work on the file questioning the legitimacy of the file.*


In [ ]:
## sshhh! Go on have a sneak-peek at the file, which is in the data folder attached to the assignment

Alignment  Intelligence  Strength  Speed  Durability  Power  \
Name                                                                      
3-D Man          good            50        31     43          32     25   
A-Bomb           good            38       100     17          80     17   
Abe Sapien       good            88        14     35          42     35   
Abin Sur         good            50        90     53          64     84   
Abomination       bad            63        80     53          90     55   

             Combat  Total  
Name                        
3-D Man          52    233  
A-Bomb           64    316  
Abe Sapien       85    299  
Abin Sur         65    406  
Abomination      95    436

**But first make necessary imports**

In [ ]:
import pandas as pd
import numpy as np

<center> <h1> Part 2: Dr. Banner Works with DataFrames </h1></center>

Fury asks Dr. Banner to help him with the file and filter out results for his army. In particular, he is interested in <i><b>Intelligence</b></i> and <b><i>Combat</i></b>, but he asks Banner to work on all the data for now.

### Q7 (2 points)

- Write a function 'read_secret_file' that takes in the filename and name of the column that will be used as the index.
- The function reads the file into a dataframe and sets the index of the dataframe based on the second argument.
- Then returns the dataframe. 
- Note that **decoded_message** gives you the name of the file but it doesn't include the file extension, .csv, or the file path (it's under the data folder that should be saved in the same directory as this notebook file). Therefore your function should first specify the filename and its path properly before attempting to open it.

Refer to the sample function call below.

In [ ]:
def read_secret_file(file_name, index_col):
    #answer goes here

In [ ]:
df_data = read_secret_file(decoded_message, "Name")

In [ ]:
df_data

Alignment  Intelligence  Strength  Speed  Durability  Power  \
Name                                                                          
3-D Man              good            50        31     43          32     25   
A-Bomb               good            38       100     17          80     17   
Abe Sapien           good            88        14     35          42     35   
Abin Sur             good            50        90     53          64     84   
Abomination           bad            63        80     53          90     55   
...                   ...           ...       ...    ...         ...    ...   
Yellowjacket         good            88        10     12          28     12   
Yellowjacket II      good            50        10     35          28     31   
Ymir                 good            50       100     27         100     83   
Zatanna              good            75        10     23          28    100   
Zoom                  bad            50        10    100          28     72   

                 Combat  Total  
Name                            
3-D Man              52    233  
A-Bomb               64    316  
Abe Sapien           85    299  
Abin Sur             65    406  
Abomination          95    436  
...                 ...    ...  
Yellowjacket         14    164  
Yellowjacket II      28    182  
Ymir                 28    388  
Zatanna              56    292  
Zoom                 28    288  

[611 rows x 8 columns]

### Q8 (2 points)
Oh, dear... Working on a secret stressful project, Banner turns into Hulk, which Fury didn't expect. <br>

Go on <i>smash</i> the dataframe.

- Write a function 'get_marvel_characters' to select Marvel Universe characters from the dataframe, mcu_characters, which is provided below.
- The function should take in the above dataframe and the mcu_characters list as parameters.
- The function should return a new dataframe containing the data about the Marvel characters provided in the list. Refer to the sample output below.

In [ ]:
#here is the list of characters to be selected
mcu_characters = [
    'Ant-Man', 'Black Panther', 'Black Widow', 'Captain America', 'Captain Marvel', 'Daredevil',
    'Deadpool', 'Doctor Strange', 'Groot', 'Iron Man', 'Hawkeye', 'Hulk', 'Loki', 'Nick Fury', 'Red Skull',
    'Rocket Raccoon', 'Spider-Man', 'Thor', 'Thanos', 'Ultron', 'Violet', 'Vision', 'War Machine', 'Wasp'
]

In [ ]:
def get_marvel_characters(df, filtering_list):
    #answer goes here

In [ ]:
df_marvel_only = get_marvel_characters(df_data, mcu_characters)
df_marvel_only

Alignment  Intelligence  Strength  Speed  Durability  Power  \
Name                                                                          
Ant-Man              good           100        10     23          28     32   
Black Panther        good            88        16     30          60     41   
Black Widow          good           100        60     90          80     89   
Captain America      good            63        19     35          56     46   
Captain Marvel       good           100       100     67          95     62   
Daredevil            good            75        13     27          28     57   
Deadpool          neutral            50        15     30         100    100   
Doctor Strange       good           100        10     12          84    100   
Groot                good            75        85     33          70     92   
Iron Man             good           100        85     58          85    100   
Hawkeye              good            50        12     23          14     26   
Hulk                 good            88       100     47         100     41   
Loki                  bad            88        57     47          85     85   
Nick Fury            good            75        11     23          42     25   
Red Skull             bad            75        10     12          14     19   
Rocket Raccoon       good            50         5     23          28     28   
Spider-Man           good            88        55     60          74     58   
Thor                 good            69       100     92         100    100   
Thanos                bad            88       100     17         100    100   
Ultron                bad            88        83     42         100    100   
Violet               good            10         4     12          28     65   
Vision               good           100        72     54          95     76   
War Machine          good            63        80     63         100    100   
Wasp                 good            63         8     58          52     23   

                 Combat  Total  
Name                            
Ant-Man              32    225  
Black Panther       100    335  
Black Widow         100    519  
Captain America     100    319  
Captain Marvel       56    480  
Daredevil            95    295  
Deadpool            100    395  
Doctor Strange       60    366  
Groot                64    419  
Iron Man             64    492  
Hawkeye              80    205  
Hulk                 85    461  
Loki                 56    418  
Nick Fury           100    276  
Red Skull            80    210  
Rocket Raccoon       64    198  
Spider-Man           85    420  
Thor                 85    546  
Thanos               80    485  
Ultron               64    477  
Violet               14    133  
Vision               70    467  
War Machine          85    491  
Wasp                 42    246

### Q9 (2 points)
Fury wants to be able to look up a character's given power. For this purpose, write a function that takes a character's name and the name of the power and returns the corresponding value from the df_marvel_only dataframe.

In [ ]:
def lookup_character_power(character, power):
    #answer goes here

In [ ]:
lookup_character_power("Black Widow", "Total")

519

### Q10 (2 points)
Write a one-line code snippet that returns the number of characters whose Alignment is bad; that is, the villains, from the df_marvel_only dataframe.

In [ ]:
#answer goes here

4

### Q11 (1 points)
Write a one-line code snippet that returns the average value of the powers among all characters in the dataframe.

In [ ]:
#answer goes here

Intelligence     76.916667
Strength         46.250000
Speed            40.750000
Durability       67.416667
Power            65.208333
Combat           73.375000
Total           369.916667
dtype: float64

### Q12 (2 points)
Fury was right! Looking at the average values, Intelligence and Combat are the strongest areas. He wants to present these data to the council. 

Write a one-line code snippet that selects the 'Alignment', 'Intelligence' and 'Combat' columns and stores them in a new dataframe, df_final.

In [ ]:
#answer goes here

In [ ]:
df_final

Alignment  Intelligence  Combat
Name                                           
Ant-Man              good           100      32
Black Panther        good            88     100
Black Widow          good           100     100
Captain America      good            63     100
Captain Marvel       good           100      56
Daredevil            good            75      95
Deadpool          neutral            50     100
Doctor Strange       good           100      60
Groot                good            75      64
Iron Man             good           100      64
Hawkeye              good            50      80
Hulk                 good            88      85
Loki                  bad            88      56
Nick Fury            good            75     100
Red Skull             bad            75      80
Rocket Raccoon       good            50      64
Spider-Man           good            88      85
Thor                 good            69      85
Thanos                bad            88      80
Ultron                bad            88      64
Violet               good            10      14
Vision               good           100      70
War Machine          good            63      85
Wasp                 good            63      42

<h3><center> 
    That's it! Fury is super happy with all the work. He goes to the council with these results. <br>
    "I still believe in heros" says Fury in conclusion.
    </center>
</h3>